server.go

```diff
@@ -1277,11 +1298,21 @@ func (s *server) handleAddPeerMsg(state *peerState, sp *serverPeer) bool {
 		delete(state.banned, host)
 	}
 
-	// TODO: Check for max peers from a single IP.
+	// Limit max number of connections from a single IP.  However, allow
+	// whitelisted inbound peers and localhost connections regardless.
+	isInboundWhitelisted := sp.isWhitelisted && sp.Inbound()
+	peerIP := sp.NA().IP
+	if cfg.MaxSameIP > 0 && !isInboundWhitelisted && !peerIP.IsLoopback() &&
+		state.ConnectionsWithIP(peerIP) > cfg.MaxSameIP {
```

Comment:
I was testing this and noticed it is off by one because the peer that is being examined is not yet added to the data structures at this point.  I'd suggest the following which accounts for the peer currently being connected:

`state.ConnectionsWithIP(peerIP)+1 > cfg.MaxSameIP {`

Alternatively, you could do the following, but I think the former reads more clearly:

`state.ConnectionsWithIP(peerIP) >= cfg.MaxSameIP {`